### PR TITLE
ci: make canary uploads less racy

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/25464

Only uploads canary git hash if the one in the bucket is present
in the current action Git history.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -486,7 +486,9 @@ jobs:
           github.ref == 'refs/heads/main')
         run: |-
           gsutil -h "Cache-Control: public, max-age=3600" cp ./target/release/*.zip gs://dl.deno.land/canary/$(git rev-parse HEAD)/
-          echo ${{ github.sha }} > canary-latest.txt
+          lastest_canary_hash=$(gsutil cat gs://dl.deno.land/canary-$(rustc -vV | sed -n "s|host: ||p")-latest.txt) && \
+          git cat-file -e $(lastest_canary_hash) && \
```
Comment (dsherret) [inline .github/workflows/ci.yml:490]: I guess this will error and fail the step? Probably ok, but might lead to main looking more red than it is?

**Review by dsherret — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml (hunk 1 above)
Comment (bartlomieju) [reply .github/workflows/ci.yml]: Improved a bit by adding a conditional

**Comment by bartlomieju** [comment]
This is a bit risky, so I'll merge it tomorrow.

**Review by tungvt-vbee — APPROVED** [review]
Good

**Pull request merged**